_posts/2024-07-23-datafusion-40.0.0.md

```diff
@@ -0,0 +1,490 @@
+
+# Building Systems is Easier with DataFusion 🛠️
+
+In addition to make small API improvements, there are several new APIs that make
+it easier to build systems on top of DataFusion, for example:
+
+* Faster and easier to use [TreeNode API] for traversing and manipulating plans and expressions.
+* All functions now use the same [Scalar User Defined Function API], making it easier to customize
+  DataFusion's behavior without sacrificing performance. See [ticket] for more details.
+* DataFusion can now be compiled to [WASM]. 
+
+[TreeNode API]: https://docs.rs/datafusion/latest/datafusion/common/tree_node/trait.TreeNode.html#overview
+[Scalar User Defined Function API]: https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.ScalarUDFImpl.html
+[ticket]: https://github.com/apache/arrow-datafusion/issues/8045
+[WASM]: https://github.com/apache/datafusion/discussions/9834
+
+## User Defined SQL Parsing Extensions
+
+As of DataFusion 40.0.0, the [`ExprPlanner`] allows easy extension of
+DataFusion's SQL planner to support custom operators or syntax.
+
+For example the [datafusion-functions-json] project uses this API add support
+for JSON operators in SQL queries. It provides a custom implementation for
+planning JSON operators such as `->` and `->>` with code like:
+
+```rust
+struct MyCustomPlanner;
+
+impl ExprPlanner for MyCustomPlanner {
+    // Provide custom implementation for planning a binary operators
+    // such as `->` and `->>`
+    fn plan_binary_op(
+        &self,
+        expr: RawBinaryExpr,
+        _schema: &DFSchema,
+    ) -> Result<PlannerResult<RawBinaryExpr>> {
+        match &expr.op {
+           BinaryOperator::Arrow => { /* plan -> operator */ }
+           BinaryOperator::LongArrow => { /* plan ->> operator */ }
+           ...
+        }
+    }
+}
+```
+
+Thanks to [@samuelcolvin], [@jayzhan211] and [@dharanad] for helping make this
+feature happen.
+
+[datafusion-functions-json]: https://github.com/datafusion-contrib/datafusion-functions-json
+[`ExprPlanner`]: https://docs.rs/datafusion/latest/datafusion/logical_expr/planner/trait.ExprPlanner.html
+
+## Pluggable Support for `CREATE FUNCTION` 
+
+DataFusion's new [`FunctionFactory`] API let's users provide a handler for
+`CREATE FUNCTION` in SQL. This feature lets you build systems that support
+defining functions in SQL such as
+
+```sql
+-- SQL based functions
+CREATE FUNCTION my_func(DOUBLE, DOUBLE) RETURNS DOUBLE
+    RETURN $1 + $3
+;
+
+-- ML Models
+CREATE FUNCTION iris(FLOAT[]) RETURNS FLOAT[] LANGUAGE TORCH AS 'models:/iris@champion';
+
+-- WebAssembly
+CREATE FUNCTION func(FLOAT[]) RETURNS FLOAT[] LANGUAGE WASM AS 'func.wasm'
+```
+
+Huge thanks to [@milenkovicm] for this feature
+
+BTW it would be great if [someone made a demo] showing how easy this is to do 🎣. 
+
+[someone made a demo]: https://github.com/apache/datafusion/issues/9326 
```

Comment:
Thanks @alamb !
One question, would it make sense to link example https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/function_factory.rs ?

Reply:
Oh, sweet, yes it would! I think I forgot about that one. Added in https://github.com/apache/datafusion-site/pull/6/commits/79e8f16ec9023328571d87250239a6061572171e